src/server/scripts/Kalimdor/TempleOfAhnQiraj/boss_huhuran.cpp

```diff
@@ -15,148 +15,147 @@
-        Wyvern_Timer = urand(18000, 28000);
-        Spit_Timer = 8000;
-        PoisonBolt_Timer = 4000;
-        NoxiousPoison_Timer = urand(10000, 20000);
-        FrenzyBack_Timer = 15000;
-
-        Frenzy = false;
-        Berserk = false;
+        _Reset();
+        berserk = false;
+        hardEnrage = false;
     }
 
-    void UpdateAI(uint32 diff) override
+    void EnterCombat(Unit* /*who*/) override
     {
-        //Return since we have no target
-        if (!UpdateVictim())
-            return;
-
-        //Frenzy_Timer
-        if (!Frenzy && Frenzy_Timer <= diff)
-        {
-            DoCast(me, SPELL_FRENZY);
-            Talk(EMOTE_FRENZY_KILL);
-            Frenzy = true;
-            PoisonBolt_Timer = 3000;
-            Frenzy_Timer = urand(25000, 35000);
-        }
-        else Frenzy_Timer -= diff;
+        events.ScheduleEvent(EVENT_FRENZY, urand(25000, 35000));
+        events.ScheduleEvent(EVENT_WYVERN_STING, urand(18000, 28000));
+        events.ScheduleEvent(EVENT_ACID_SPIT, 8000);
+        events.ScheduleEvent(EVENT_NOXIOUS_POISON, urand(10000, 20000));
+        events.ScheduleEvent(EVENT_HARD_ENRAGE, 300000);
+    }
 
-        // Wyvern Timer
-        if (Wyvern_Timer <= diff)
-        {
-            DoCastAOE(SPELL_WYVERNSTING);
-            Wyvern_Timer = urand(15000, 32000);
-        }
-        else Wyvern_Timer -= diff;
+    void JustDied(Unit* /*killer*/) override
+    {
+        _JustDied();
+    }
 
-        //Spit Timer
-        if (Spit_Timer <= diff)
+    void DamageTaken(Unit*, uint32& /*damage*/, DamageEffectType, SpellSchoolMask) override
+    {
+        if (!berserk && HealthBelowPct(30))
         {
-            DoCastVictim(SPELL_ACIDSPIT);
-            Spit_Timer = urand(5000, 10000);
+            DoCastSelf(SPELL_BERSERK, true);
+            Talk(EMOTE_BERSERK);
+            berserk = true;
         }
-        else Spit_Timer -= diff;
+    }
 
-        //NoxiousPoison_Timer
-        if (NoxiousPoison_Timer <= diff)
-        {
-            DoCastVictim(SPELL_NOXIOUSPOISON);
-            NoxiousPoison_Timer = urand(12000, 24000);
-        }
-        else NoxiousPoison_Timer -= diff;
+    void UpdateAI(uint32 diff) override
+    {
+        if (!UpdateVictim())
+            return;
 
-        //PoisonBolt only if frenzy or berserk
-        if (Frenzy || Berserk)
+        events.Update(diff);
+        while (uint32 eventid = events.ExecuteEvent())
         {
-            if (PoisonBolt_Timer <= diff)
+            switch (eventid)
             {
-                DoCastVictim(SPELL_POISONBOLT);
-                PoisonBolt_Timer = 3000;
+                case EVENT_FRENZY:
+                    DoCastSelf(SPELL_FRENZY, true);
+                    Talk(EMOTE_FRENZY_KILL);
+                    events.RepeatEvent(urand(25000, 35000));
+                    break;
+                case EVENT_WYVERN_STING:
+                    if (Unit* target = SelectTarget(SelectTargetMethod::MinDistance))
```

Comment:
It's an AoE spell, you will need to filter targets in the spell script so only the closest ones are selected

Reply:
idk about ::FilterTargets based on distance so I'll leave it with limited targets to at least have the target amount right.
Open for distancePriorityCheck after merge or commit in the branch if you want to do it.

Reply:
This is for learning purposes only: expliciting a target to an AoE spell bears no effect because the target selector will ignore it (the spell itself has no target so we cannot specify one, it will still follow the default selector for the spell, which in our case is TARGET_UNIT_SRC_AREA_ENEMY)